Refactor Geometry classes to subclass the C extension type

jorisvandenbossche opened the pull request:
.. and use the pygeos functions for all the Geometry methods/properties. 

This is a first step for integrating PyGEOS into Shapely (https://github.com/Toblerity/Shapely/issues/962):

- For now I still use PyGEOS as an external dependency (it will be merged into Shapely in a later step, I first wanted to ensure this essential part is working out)
- This makes all the Geometry classes (Point, LineString, Polygon, etc) subclasses of the pygeos.Geometry extension type.
- This also registers those subclasses to pygeos. This means that all pygeos functions will start to return the correct shapely subclass, and not the generic pygeos.Geometry
- For most of the methods and properties of the Geometry classes, I switched to using pygeos functions (eg ` geom_factory(self.impl['difference'](self, other))` to `pygeos.difference(self, other)`). For other functionality in most other modules (eg shapely.ops etc), this is still left as is and keeps using the ctypes interface (this can be cleaned up later in a separate PR)
- A large chunk of the diff is actually the re-implementation of the Point(), Linestring(..), Polygon(..) etc constructors, as those don't directly map to the pygeos functions to create geometries from coordinates (the shapely ones are more flexible / allow more variation in input types)
- All tests are passing locally (with pygeos master), but I xfailed a bunch of tests where there are small differences or where we still need to implement something in the pygeos code to support it. But so it can be seen explicitly in the diff which ones are not yet passing.

cc @caspervdw

jorisvandenbossche commented on shapely/coords.py:
```diff
@@ -37,121 +37,67 @@ class CoordinateSequence(object):
-        self._ndim = self.__p__._ndim
-        self._cseq = lgeos.GEOSGeom_getCoordSeq(self.__p__._geom)
+    def __init__(self, coords):
+        self._coords = coords
```
What exactly to do with this `CoordinateSequence` class is something still to be discussed, and I opened https://github.com/Toblerity/Shapely/issues/984 for that

caspervdw replied:
I have been reading up on https://github.com/Toblerity/Shapely/issues/984 and I tend to go with this implementation for this PR (making `self._coords` a ndarray directly)

However I think that sticking a full duplicate of `coords` to the object by default is a bit too much. Also `_coords` won't be initialized properly in all construction paths.

Within this approach (I actually like the lazy CoordinateSequency object approach better) I think it is best to use a pattern like this:

```
@property
def coords(self):
    try:
        coords = getattr(self, "_coords")
    except AttributeError:
        coords = pygeos.get_coordinates(self)
        coords.flags.writeable = False
        self._coords = coords
    return coords
```

jorisvandenbossche replied:
> However I think that sticking a full duplicate of coords to the object by default is a bit too much. 

Note that this CoordinateSequence is not constructed by default when creating a geometry object. It will only be created once a user accesses the `geom.coords` attribute. 

Now, with the current code, each time the user does `geom.coords`, it will create a new CoordinateSequence object containing a new array, so making this "cached" is certainly a good idea (as well as making the array not writable)

jorisvandenbossche commented on shapely/geometry/collection.py:
```diff
@@ -35,11 +37,18 @@ def __init__(self, geoms=None):
-        else:
-            self._geom, self._ndim = geos_geometrycollection_from_py(geoms)
+            # TODO better empty constructor
+            return pygeos.from_wkt("GEOMETRYCOLLECTION EMPTY")
```
For all those constructors, I now use WKT to create an empty geometry, but see https://github.com/pygeos/pygeos/issues/149 to discuss if we need to have a better way for this

jorisvandenbossche commented on shapely/geometry/point.py:
```diff
@@ -44,35 +46,65 @@ def __init__(self, *args):
     # Coordinate getters and setters
 
+    def __repr__(self):
+        return "<shapely.geometry.Point {} >".format(self.wkt)
```
TODO need to move this to the base class

jorisvandenbossche commented on tests/test_create_inconsistent_dimensionality.py:
```diff
@@ -38,16 +40,18 @@
     with pytest.raises(ValueError) as exc:
         wkt = shape(geojson).wkt
-    assert exc.match("Inconsistent coordinate dimensionality")
+    assert exc.match("Inconsistent coordinate dimensionality|Input operand 0 does not have enough dimensions")
```
We get different error message in several places (we should see to what extent we want to preserve them, sometimes they are more informative though)

jorisvandenbossche commented on tests/test_geos_err_handler.py:
```diff
@@ -2,11 +2,16 @@
+from tests.conftest import shapely20_todo
+
 
+@shapely20_todo  # logging is not yet implemented
```
This is something we need to check in pygeos how this can be possible, if it's a feature we want to preserve

mwtoews replied:
I'd say all error messages from GEOS (like the one below) should be converted to exception messages, similar to https://github.com/Toblerity/Shapely/issues/991#issuecomment-699147566 . I'd assume it would be difficult to capture an error message in pygeos, but also emit the same error message to the logger handler below. I wouldn't expect many other folks to be relying on these log capture methods.

jorisvandenbossche replied:
Opened https://github.com/Toblerity/Shapely/issues/998 for a discussion on logging.

jorisvandenbossche commented on tests/test_persist.py:
```diff
@@ -16,6 +18,7 @@ def test_pickle(self):
         q = pickle.loads(data)
         self.assertTrue(q.equals(p))
 
+    @shapely20_todo  # big_endian is not yet implemented
```
In pygeos right now we didn't implement this keyword, but have a `byte_order` keyword instead (matches more closely the terminology of GEOS). The default is to use the machine's endianness, and one can use the keyword to override this.

jorisvandenbossche commented on tests/test_pickle.py:
```diff
@@ -19,6 +19,9 @@ def test_pickle_round_trip(cls, coords):
     data = dumps(geom1, HIGHEST_PROTOCOL)
     geom2 = loads(data)
     assert geom2.has_z == geom1.has_z
+    # TODO(shapely-2.0) LinearRing type not preserved in roundtrip
```
We were just discussing this in PyGEOS as well -> https://github.com/pygeos/pygeos/pull/190 

There we thought that it might be OK enough to get a LineString back from pickling a LinearRing. However, having the pickling on the python side (still using WKB), we could more easily track the original class as well and ensure it is correctly preserved. 
However, having it in python compared to in the base C extension type has a performance penalty.

jorisvandenbossche replied:
This is tracked by https://github.com/Toblerity/Shapely/issues/988

jorisvandenbossche commented on tests/test_predicates.py:
```diff
@@ -43,7 +46,7 @@ def test_binary_predicate_exceptions(self):
               (459, 415), (399, 381), (519, 311), (520, 242), (519, 173),
               (399, 450), (339, 207)]
-        self.assertRaises(TopologicalError, Polygon(p1).within, Polygon(p2))
+        self.assertRaises(pygeos.GEOSException, Polygon(p1).within, Polygon(p2))
```
Currently, shapely has a bunch of custom error classes in https://github.com/Toblerity/Shapely/blob/master/shapely/errors.py. But in PYGEOS, for now we just have a single GEOSException for errors that are directly raised from one of the GEOS functions.

sgillies replied:
Raising pygeos exceptions is just an intermediate state of the API then?

jorisvandenbossche replied:
Yes, the fact that it is a "pygeos" exception is intermediate state. But basically everywhere you see `pygeos` you need to mentally replace it with `shapely` (as in one of the next steps we will move pygeos into shapely and replace the pygeos imports with internal shapely imports). 

So the "real" question here is if we are, long term, fine with having only a generic `shapely.GEOSException`, or if we want to have more fine grained error *classes* (TopologicalError, WKBReadingError, WKTReadingError). 
Note that this is only about the class of the exception, also the generic GEOSException error class will still have the custom, informative error messages.

jorisvandenbossche replied:
Opened https://github.com/Toblerity/Shapely/issues/991 to have a dedicated place to discuss this topic.

jorisvandenbossche commented:
Gentle ping to get some feedback here.

mwtoews commented:
Thanks @jorisvandenbossche this looks great so far. I've built this PR locally, and the majority of the tests indeed pass. I'll drop a few review comments when I get spare moments...

sgillies commented:
I won't be able to try this out until the weekend. Will report back then.

caspervdw commented on shapely/geometry/base.py:
```diff
@@ -164,9 +152,7 @@ class BaseGeometry(object):
     #     case of a multipart geometry member.
-    __geom__ = EMPTY
     __p__ = None
-    _ndim = None
```
I think that best all class attributes should be dropped in shapely 2.0. I read these as defaults for object-level attributes. These attributes add unnecessary state to the geometry objects.

As they are all private, I think they can be dropped. `_crs` and `_is_empty` could be replaced by properties calling `pygeos.get_srid` and `pygeos.is_empty` respectively if that is convenient for the implementation.

jorisvandenbossche replied:
Indeed, this needs further clean-up. I only removed the ones that I encountered explicitly while refactoring for now, but normally all others can be removed as well (they shouldn't be used anymore, or at least eventually they won't be used anymore). But so for sure, none of those needs to stay.

caspervdw commented on shapely/geometry/base.py:
```diff
@@ -60,34 +62,20 @@ def dump_coords(geom):
-        ob._ndim = 2
-    ob._is_empty = False
-    return ob
+    ob = _DummyGeometry(g)
```
This seems backwards to me.
In my opinion we should never pass these pointers around as python `int` objects. Do you think we can remove the `geom_factory` completely from shapely 2.0?

jorisvandenbossche replied:
Yes, this is certainly only temporary. I don't refactor the full of Shapely in a single PR, and thus in other parts of shapely, `geom_factory` is still used together with ctypes. So those changes here with the DummyGeometry hack is just to ensure `geom_factory` keeps working. But the end goal of Shapely 2.0 for sure means that all usage of `geom_factory` is eliminated.

caspervdw reviewed (commented):
@jorisvandenbossche  Thanks for this enormous effort of combining pygeos and shapely. I left some remarks.

jorisvandenbossche commented on shapely/geometry/linestring.py:
```diff
@@ -39,11 +41,37 @@ def __init__(self, coordinates=None):
+            return pygeos.from_wkt("LINESTRING EMPTY")
+        elif isinstance(coordinates, LineString):
+            if type(coordinates) == LineString:
+                # TODO should we clone it?
```
Just checked, eg copying a tuple (an immutable object) also returns the same object. So that seems like a common pattern. Will remove the comment then, and simply return the original object.

jorisvandenbossche commented on shapely/geometry/polygon.py:
```diff
@@ -176,9 +206,8 @@ class Polygon(BaseGeometry):
 
     _exterior = None
     _interiors = []
```
I am not sure what this `_interiors` attribute was used for originally, but in any case it's not used anymore now, so will simply remove.

jorisvandenbossche commented:
@caspervdw Thanks for the review!

jorisvandenbossche commented:
@caspervdw a build related question: pygeos doesn't seem to picking up the correct GEOS version (from travis ci output: *"ImportError: The shapely GEOS version (3.7.3-CAPI-1.11.3 b50468f) is incompatible with the GEOS version PyGEOS was compiled with (3.5.0-CAPI-1.9.0)"*) 
I am actually not fully sure where it is getting this GEOS version, but in any case there are multiple GEOS versions installed (and cached) on the travis image. I prepended the one to be used to the PATH, but doesn't seem to have fixed it.

jorisvandenbossche commented:
Note: the tests on Travis are now passing for the Python 3.8 build. For older versions only one MultiLineString.interpolate test is failing that I can't directly reproduce locally.

caspervdw commented on shapely/geometry/polygon.py:
```diff
@@ -72,13 +109,17 @@ def __setstate__(self, state):
     @property
     def is_ccw(self):
         """True is the ring is oriented counter clock-wise"""
+        # TODO
```
See https://github.com/pygeos/pygeos/pull/201

mwtoews commented on shapely/geometry/base.py:
```diff
@@ -838,32 +765,19 @@ class GeometrySequence(object):
     _ndim = None
```
`__p__` is used below, but I think this would be a good time to rename this private variable to `_parent`

jorisvandenbossche replied:
Specifically for this `GeometrySequence` class, we also still need to see what to do with it (keep it or not?)

mwtoews commented on shapely/geometry/base.py:
```diff
@@ -154,75 +142,42 @@ class BaseGeometry(object):
-        self.__geom__ = val
+    @property
+    def __geom__(self):
+        return self._ptr
```
`__geom__` is another ["dunder"](https://docs.python.org/3/reference/lexical_analysis.html#reserved-classes-of-identifiers) I think we could remove for shapely-2.0, as it's not a system-defined name. I see that this holds the same value as `_geom` and `_ptr` (from pygeos).

jorisvandenbossche replied:
Yes, the attributes giving read access to the pointer (`_geom`, `__geom__`, and now also `_ptr`) can use a clean-up. But for now, those are kept here in any case for compatibility with the rest of shapely that is not yet adapted.

So we should probably decide which one to keep (eg `_geom`) and deprecate the other already in Shapely 1.8 (eg `__geom__`)

(might also be worth a separate issue, as it is not "solvable" for this PR)

jorisvandenbossche replied:
Opened https://github.com/Toblerity/Shapely/issues/999 for this `__geom__` / `_geom` dicussion

mwtoews commented on shapely/ops.py:
```diff
@@ -121,37 +122,15 @@ def cascaded_union(self, geoms):
-            subs[i] = g._geom
-        collection = lgeos.GEOSGeom_createCollection(6, subs, L)
-        return geom_factory(lgeos.methods['cascaded_union'](collection))
+        return pygeos.union_all(geoms, axis=None)
```
Is pygeos calling (deprecated) `GEOSUnionCascaded_r`? This is the same return for `unary_union` below.

jorisvandenbossche replied:
> Is pygeos calling (deprecated) GEOSUnionCascaded_r? This is the same return for unary_union below.

No, it's actually calling GEOSUnaryUnion_r, and not GEOSUnionCascaded_r (GEOSUnionCascaded_r is simply not implemented in pygeos). Note that this is similar to what Shapely does: for GEOS >= 3.3, `lgeos.methods['cascaded_union']` is actually an alias of `unary_union`, only for older GEOS versions it is using GEOSUnionCascaded (but pygeos doesn't support such old GEOS version):

https://github.com/Toblerity/Shapely/blob/73626bce8e99b3f251ad874a41d97d2384e734ce/shapely/geos.py#L838

Two potential discussion items out of this:

- What's the minimal supported GEOS version we want for Shapely 2.0? Shapely starts testing for GEOS 3.5 and up (the same for pygeos), but actually has code in shapely/geos.py going back to GEOS 3.0

- Do we actually want to deprecate the python `cascaded_union` function, instead of keeping it around as alias of `unary_union`? (although this is quite independent from all shapely 2.0 / pygeos refactors, but the deprecations in 1.8 could be an opportunity to deprecate this as well, if we want. But I have the impression it is still used quite a lot)

(again might be worth turning into dedicated issues)

jorisvandenbossche replied:
Opened https://github.com/Toblerity/Shapely/issues/1000 for the minimum supported GEOS version and https://github.com/Toblerity/Shapely/issues/1001 for the unary_union/cascaded_union aliasing

mwtoews commented on shapely/ops.py:
```diff
@@ -121,37 +122,15 @@ def cascaded_union(self, geoms):
-            subs[i] = g._geom
-        collection = lgeos.GEOSGeom_createCollection(6, subs, L)
-        return geom_factory(lgeos.methods['unary_union'](collection))
+        return pygeos.union_all(geoms, axis=None)
```
similar to comment above, is pygeos calling `GEOSUnaryUnion_r` here?

jorisvandenbossche replied:
> similar to comment above, is pygeos calling GEOSUnaryUnion_r here?

Yes, it's a potential naming issue on the pygeos side, but `union_all` is basically `unary_union` (the main difference being that it supports an `axis` keyword in case you have a multidimensional array to union along a specific axis). But we should certainly consider addng a `unary_union` function as well to be compatible with Shapely's current naming.

jorisvandenbossche replied:
Relevant issue about `union_all` vs `unary_union` is https://github.com/pygeos/pygeos/issues/194

mwtoews commented on tests/test_point.py:
```diff
@@ -139,7 +142,9 @@ def test_coords(self):
-    assert p.bounds == ()
+    # TODO keep this change or not?
+    assert len(p.bounds) == 4
+    assert all(math.isnan(v) for v in p.bounds)
```
Why does just `point empty` have four nan bounds, and not other empty geometry types (that return an empty tuple `()`)? Otherwise, yes this idea is worth considering.

jorisvandenbossche replied:
Ah, that's an oversight / something to clean-up. For the base class method, I left the current behaviour but added a TODO comment:

```diff
     @property
     def bounds(self):
         """Returns minimum bounding region (minx, miny, maxx, maxy)"""
+        # TODO return empty tuple or (nan, nan, nan, nan)
         if self.is_empty:
             return ()
         else:
-            return self.impl['bounds'](self)
+            return tuple(pygeos.bounds(self).tolist())
```

while this property is overridden by Point, where I actually removed this empty tuple return:

```diff
    @property
    def bounds(self):
        """Returns minimum bounding region (minx, miny, maxx, maxy)"""
-        try:
-            xy = self.coords[0]
-        except IndexError:
-            return ()
-        return (xy[0], xy[1], xy[0], xy[1])
+        # TODO keep return type of tuple, or just numpy array?
+        return tuple(pygeos.bounds(self).tolist())
```

So for sure, we need to be consistent about this, whatever decision we make. 

(and actually, with the new implementation, we can remove the override in the Point subclass, as the base class implementation should work fine for Points as well)

jorisvandenbossche replied:
Removed the override in `Point` class, and leaving the discussion to changy `empty.bounds` for a separate issue

jorisvandenbossche replied:
Opened https://github.com/Toblerity/Shapely/issues/1023 to discuss this

jorisvandenbossche commented:
@mwtoews thanks a lot for the review / questions !

mwtoews commented:
> Note: the tests on Travis are now passing for the Python 3.8 build. For older versions only one MultiLineString.interpolate test is failing that I can't directly reproduce locally.

I can reproduce this locally with GEOS 3.7.1:
```python
from shapely import wkt
multiline = wkt.loads('MULTILINESTRING ((0 0, 2 0), (3 0, 3 6))')
multiline.interpolate(0).wkt

import pygeos
multiline = pygeos.Geometry('MULTILINESTRING ((0 0, 2 0), (3 0, 3 6))')
pygeos.line_interpolate_point(multiline, 0)
```
this should return `POINT(0 0)`, but pygeos returns `POINT EMPTY`. See [src/ufuncs.c](https://github.com/pygeos/pygeos/blame/7a2798ab8ef43acf59e2023b8cfcb49e6d9384d3/src/ufuncs.c#L341-L365) for the conditional implementation.

I'm not sure how to describe this issue yet, but I'm sure @caspervdw could shed some light. It seems that the logic does not support linear referencing with MutiLineStrings?

jorisvandenbossche commented:
@mwtoews thanks for the catch!

mwtoews commented on shapely/geometry/base.py:
```diff
@@ -748,11 +679,8 @@ def interpolate(self, distance, normalized=False):
-            op = self.impl['interpolate']
-        return geom_factory(op(self, distance))
+        # TODO spelling normalize(d)
+        return pygeos.line_interpolate_point(self, distance, normalize=normalized)
```
See https://github.com/pygeos/pygeos/pull/209

jorisvandenbossche commented:
With the upstream fix in pygeos for `line_interpolate_point` (thanks @caspervdw!), Travis is green now ;)

jorisvandenbossche commented:
Appveyor is also passing, except for 1 test in the Python 3.8 build:

```
================================== FAILURES ===================================
___________________________ test_pickle_persistence ___________________________
    @requires_geos_342
    def test_pickle_persistence():
        """
        Don't crash trying to use unpickled GEOS handle.
        """
        tree = STRtree([Point(i, i).buffer(0.1) for i in range(3)])
        pickled_strtree = pickle.dumps(tree)
        unpickle_script_file_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "unpickle-strtree.py")
        proc = subprocess.Popen(
            [sys.executable, str(unpickle_script_file_path)],
            stdin=subprocess.PIPE,
        )
        proc.communicate(input=pickled_strtree)
        proc.wait()
>       assert proc.returncode == 0
E       assert 1 == 0
E        +  where 1 = <subprocess.Popen object at 0x000001F9904BD520>.returncode
tests\test_strtree.py:102: AssertionError
---------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "C:\projects\shapely\tests\unpickle-strtree.py", line 8, in <module>
    from shapely.geometry import Point
  File "C:\projects\shapely\shapely\geometry\__init__.py", line 4, in <module>
    from .base import CAP_STYLE, JOIN_STYLE
  File "C:\projects\shapely\shapely\geometry\base.py", line 18, in <module>
    import pygeos
  File "c:\python38-x64\lib\site-packages\pygeos\__init__.py", line 1, in <module>
    from .lib import GEOSException  # NOQA
ImportError: DLL load failed while importing lib: The specified module could not be found.
```

I am not very familiar with such Windows issues, but so within subprocess call, it is not finding the GEOS library.

mwtoews commented:
I've been checking the performance difference with this PR (with pygeos master) and the current shapely master, both with libgeos master '3.8.0'. These were tested in Python 3.7.7. Here is an excerpt of my findings:
```
from shapely.geometry import Point, Polygon, MultiLineString
from shapely.ops import linemerge
from shapely import wkt

# this PR
%timeit Point(1, 2)
13.6 µs ± 355 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
# shapely master
%timeit Point(1, 2)
5.78 µs ± 67.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

# this PR
%timeit wkt.loads('LINESTRING(2.4 4.1, 2.4 3, 3 3)')
4.66 µs ± 182 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
# shapely master
%timeit wkt.loads('LINESTRING(2.4 4.1, 2.4 3, 3 3)')
10.9 µs ± 348 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

# this PR
%timeit Point(1, 2).buffer(3)
73.6 µs ± 670 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
# shapely master
%timeit Point(1, 2).buffer(3)
49.7 µs ± 1.74 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

p1 = Point(0.0, 0.0)
p2 = Point(-1.0, -1.0)
# this PR
%timeit p1.disjoint(p2)
6.49 µs ± 63.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
# shapely master
%timeit p1.disjoint(p2)
1.86 µs ± 36.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

lines = MultiLineString([((0, 0), (1, 1)), ((2, 0), (2, 1), (1, 1))])
# this PR
%timeit linemerge(lines)
22.7 µs ± 1.28 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
# shapely master
%timeit linemerge(lines)
9.6 µs ± 368 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

poly = Polygon([(0,1), (1, 2), (2, 1), (1, 0), (0, 1)])
# this PR
%timeit poly.minimum_rotated_rectangle
636 µs ± 21.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
# shapely master
%timeit poly.minimum_rotated_rectangle
289 µs ± 13 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
I'm generally seeing a bit of a slowdown with pygeos compared to ctypes used with shapely master. Any ideas on why this is? Should we be concerned? (Also, did I do these tests correctly, or is there something funny on my end?)

jorisvandenbossche commented:
@mwtoews thanks a lot for checking this!

I profiled some of those cases, and some observations:

- The `Point()` constructor is slower because of the overhead in `_wrap_construct_ufunc`, which wraps the input numbers/arrays into a single array of the appropriate shape (https://github.com/pygeos/pygeos/blob/e475152600cd356130264c44723961a19a2d7280/pygeos/creation.py#L19-L27). For a single Point, this gives a large (*relative*)  overhead, though. 
  Using the directly exposed ufunc instead of the more user-friendly wrapper, and passing it the 2-element array, you can see the difference:

    ```
    In [7]: %timeit pygeos.points(1, 2)
    14.9 µs ± 2.61 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)
    
    In [11]: %timeit pygeos.lib.points(np.array([1, 2]))
    2.17 µs ± 114 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
    ```

  When working with larger arrays, the overhead of the wrapper is minimal, but to construct a single Point, this overhead becomes significant. So the `points()` function is simply not really optimized for this case of creating a single point. 
  If we want this, I think it is certainly possible to write a "single-point constructor" that optimizes for this use case.

* `wkt.loads` is already faster with this branch

* The buffer operation (`Point(1, 2).buffer(3)`, but only profiling the actual buffer) -> it seems that here it's the actual call into C that takes longer. So for that would need to take a closer look / use a different profiler. 

* The `p1.disjoint(p2)` case: also here the call into C is actually already slower as Shapely (eg timing `%timeit pygeos.lib.disjoint(p1, p2)` instead of the wrapper function). I suppose that here the ufunc machinery gives some overhead (also in C, it's not a plain call to GEOS, but done inside a inner loop of the ufunc. For scalars (0-dim arrays), this will always give *some* overhead. 
  In addition, some of the overhead of the python function also comes from the `multithreading_enabled` decorator. Again, this overhead is insignificant when working with arrays, but not for scalar geometries.

* `linemerge` -> this implementation is not yet changed (so I suppose the slowdown we see comes from other APIs being used inside Shapely's linemerge function)

* `poly.minimum_rotated_rectangle` -> this implementation relies on doing an `affine_transform`, and this is also something that is not yet updated to use pygeos (I would need to take a look into `affinity.py` first to be sure whether the use of pygeos will be able to make this faster, but I suppose we can use the vectorized transform functionality already included in pygeos to apply the transform on all coordinates of a Geometry, instead of the iteration through all parts and all points in python as is currently done in affinity.py's `affine_transform`. So this should be able to speed-up this functionality).
  In addition, GEOS actually has a `GEOSMinimumRotatedRectangle` since GEOS 3.6, so that is something we should add bindings for and can use that if available instead of our own implementation.

**In summary**:

So in general: yes, there is a slowdown for methods/properties of scalar Geometry objects. 
For some things (like the constructors of the different geometry classes), we can optimize this by spending some effort in having an efficient single-geometry version (this can probably be done in the class constructor itself, eg by using `pygeos.lib.points` instead of `pygeos.points`). 
But for the bulk of the ufuncs (the spatial predicates, operations, etc. I.e. those that directly wrap one of the `GEOS<..op..>_r` functions), this is the consequence of having general ufunc (array-aware) functions. This ufunc machinery will always have *some* overhead, and for single geometries this overhead can be noticeable (and seems to be larger as the benefit from not using ctypes).

But the question is then: **do we care about this small overhead for working with single geometries?** 
And to be clear, I care about performance (and where *easily* possible, we should fix this overhead). But 1) it is still about very small *absolute* numbers (eg for the `disjoint` method from 2 -> 8 µs), and 2) such a slowdown typically only becomes a problem if you are doing it *many times*. And that is exactly where pygeos/Shapely 2.0 will provide a big benefit with the array APIs (but with the caveat: the user will need to rewrite their loops to make use of the vectorized functions to get this benefit).

mwtoews commented:
Thanks for the thorough followup @jorisvandenbossche ! I should have picked some of these examples a bit more carefully, as indeed some are not implemented outside shapely.

I was generally suspecting there was a bit of additional overhead with ufuncs, which is expected as that's the point of pygeos. And obviously there are incredible performance gains when these constructors, operators (etc.) are used in a vectorized fashion (i.e. examples from https://caspervdw.github.io/Introducing-Pygeos/ are still valid).

Do we care about small overhead for working with single geometries? Hopefully not. Before GeoPandas, a typical workflow that required processing multiple geometries would use for-loops and lists of geometries. However, where performance is important, folks should now be migrating their workflow to use vectorized approaches.

As for specific cases of slowdowns, we can evaluate these as needed (if at all). Single geometry constructors could perhaps be investigated later. The priority for this PR is to get the expected results, which progressing really well!

Pending tasks are to look into the Windows AppVeyor fail (I could try), and I'd like to look into the `@shapely20_todo` marked tests a bit more. So hopefully another week or two??

jorisvandenbossche commented on tests/test_linear_referencing.py:
```diff
@@ -57,6 +57,7 @@ def test_line2_interpolate(self):
 
     @unittest.skipIf(geos_version < (3, 2, 0), 'GEOS 3.2.0 required')
     def test_multiline_interpolate(self):
+        print(self.multiline.interpolate(0.5))
```
Need to clean-up

jorisvandenbossche commented:
> Pending tasks are to look into the Windows AppVeyor fail (I could try)

Yes, if someone could help looking into that, that would be very helpful! 

> and I'd like to look into the @shapely20_todo marked tests a bit more.

Listing them here to get an easier overview:

* `tests/test_create_inconsistent_dimensionality.py::test_create_from_wkt`

  This is about WKT parsing returning 2 or 3 dimensions when parsing an inconsistent string. The strange thing is that I can't reproduce the difference between shapely and pygeos in an *interactive console*:

    ```python
    from shapely import wkt 
    wkt_geom = 'LINESTRING (1 1 1, 2 2)' 
    geom = wkt.loads(wkt_geom)
    print(geom.wkt)
    ```

  The snippet above extracted from the test returns for me `'LINESTRING Z (1 1 1, 2 2 0)'` in an interactive console (even in an interactive console with an older, released Shapely version), but when running it the test, it gives `'LINESTRING (1.0000000000000000 1.0000000000000000, 2.0000000000000000 2.0000000000000000)'`. 
  Is there something in the test setup that might be influencing this? (also the trimming of trailing zeros is different ...)

* `tests/test_geos_err_handler.py::test_error_handler_exception/test_info_handler`

  I had already opened https://github.com/Toblerity/Shapely/issues/998 to further discuss the topic of logging.

* `tests/test_geos_err_handler.py::test_error_handler_for_bytes`

  Currently, PyGEOS accepts both strings as bytes in `from_wkt` / `wkt.loads`, while there is a test to check that passing bytes raises an error. But it is maybe fine to change this test and accept bytes as well as pygeos is doing?

* `tests/test_persist.py::test_wkb_dumps_endianness`

  Shapely has a `big_endian=True/False` keyword in ``wkb.dump(s)``, while pygeos uses the `byte_order=0/1` matching GEOS' terminology. 
  For now I skipped the tests, but it should actually be easy to adapt and forward the `big_endian` keyword as `byte_order` to the pygeos method. 
  The only remaining question long term is then if we keep it as `big_endian` keyword in Shapely, or add the `byte_order` as alias, or eventually replace it with `byte_order`?

* ``tests/test_singularity.py::PolygonTestCase::test_polygon_3``

  This case tests the creation of ` Polygon([(1,1), (1,1), (1,1)])`, something that fails with pygeos (complaining about needing 4 coordinates to create a linearring, but when the coordinates are not all equal, this is not required). 
  I think this is something we can fix in the pygeos C code, opened https://github.com/pygeos/pygeos/issues/213 for this.

Apart from those, there are a few tests that changed (eg to correct the WKT of empty geometries), and some error messages / exception class changes (and for that we have https://github.com/Toblerity/Shapely/issues/991 to discuss this)

jorisvandenbossche commented:
Any further comments on this? 
I would like to get this merged, so I can more easily work on follow-up PRs. In my comment above, there is a list of "skipped" tests, I think it would be fine to handle those in follow-ups (I can make issues for them to ensure we don't forget it).

Further, there is the Appveyor failure (see https://github.com/Toblerity/Shapely/pull/983#issuecomment-707706532). I don't have an easy way to reproduce/investigate this. But also, since it's about calling from subprocess, it's maybe also not a blocker for this PR.

mwtoews commented:
> Listing them here to get an easier overview:

Thanks, most of these are fine to mark as TODO items for now. Here are a few specific responses:

> * `tests/test_create_inconsistent_dimensionality.py::test_create_from_wkt`
>   ... The strange thing is that I can't reproduce the difference between shapely and pygeos in an _interactive console_

See https://github.com/Toblerity/Shapely/blob/86ad1031cbcceaaa615092c1af4c55b7c8efc820/tests/__init__.py#L27-L28 which resets the defaults for GEOS' WKTWriter, which has defaults `trim = False` and `rounding_precision = -1`. That's why pytest behaves different than an interactive console.

> * `tests/test_geos_err_handler.py::test_error_handler_for_bytes`
>   Currently, PyGEOS accepts both strings as bytes in `from_wkt` / `wkt.loads`, while there is a test to check that passing bytes raises an error. But it is maybe fine to change this test and accept bytes as well as pygeos is doing?

This is fine with me.

> * `tests/test_persist.py::test_wkb_dumps_endianness`
>   Shapely has a `big_endian=True/False` keyword in `wkb.dump(s)`, while pygeos uses the `byte_order=0/1` matching GEOS' terminology.
>   For now I skipped the tests, but it should actually be easy to adapt and forward the `big_endian` keyword as `byte_order` to the pygeos method.
>   The only remaining question long term is then if we keep it as `big_endian` keyword in Shapely, or add the `byte_order` as alias, or eventually replace it with `byte_order`?

I made this call while implementing WKTWriter for shapely, as the meaning of this boolean property `big_endian=True/False` is more clear than `byte_order=0/1`. However, I don't have any objections if this is made an alias to align closer to GEOS' terminology.

mwtoews commented:
> Further, there is the Appveyor failure (see [#983 (comment)](https://github.com/Toblerity/Shapely/pull/983#issuecomment-707706532)). I don't have an easy way to reproduce/investigate this. But also, since it's about calling from subprocess, it's maybe also not a blocker for this PR.

I debugged this directly on the AppVeyor VM, and this is an odd one. It is only an issue with Python 3.8.0 (both x86 and x64), and the issue happens when "import pygeos" is attempted outside the shapely pytest environment (including subprocess). I can break the shapely pytest environment by importing pygeos before shapely. However, I'll agree that this is not a blocker for now.

mwtoews reviewed (approved):
Thanks for the huge effort @jorisvandenbossche and pygeos team! I'll let you merge this one, as you may want to squash and merge with a concise commit message.

jorisvandenbossche commented:
> See https://github.com/Toblerity/Shapely/blob/86ad1031cbcceaaa615092c1af4c55b7c8efc820/tests/__init__.py#L27-L28 which resets the defaults for GEOS' WKTWriter, which has defaults `trim = False` and `rounding_precision = -1`. That's why pytest behaves different than in an interactive console.

Ah, thanks, that clarifies ;)
Would it be fine to remove that? The comment indicates this was done for being able to pass the tests for all GEOS versions, but it seems only for GEOS versions > 3.3, and now we don't test anymore with older versions.

OK, will update the few small remaining items that can easily be done here (fixing the WKT tests, endianness, etc), and then open follow-up issues for the other remaining items.

mwtoews commented:
> Would it be fine to remove that? The comment indicates this was done for being able to pass the tests for all GEOS versions, but it seems only for GEOS versions > 3.3, and now we don't test anymore with older versions.

This was an old hack to run the tests with the older library and compare similar WKT, so yes we can update the WKT and remove it. I can volunteer to do this.

jorisvandenbossche commented:
Skipped the problematic tests on Appveyor, and updated the WKT/WKB related tests (to no longer be skipped).

> I made this call while implementing WKTWriter for shapely, as the meaning of this boolean property big_endian=True/False is more clear than byte_order=0/1. However, I don't have any objections if this is made an alias to align closer to GEOS' terminology.

For now I added `big_endian=True/False` back as alias for byte_order, we can later discuss if we want to keep both as aliases or choose one long term.

jorisvandenbossche commented:
Indeed, already removed it ;)

jorisvandenbossche commented:
OK, going to merge this, so we can further work in separate follow-up PRs. Additional comments on the changes here are of course still welcome as well, and I will open issues to track the currently remaining discussion items. Thanks all for the review!

jorisvandenbossche commented:
I created some more follow-up issues based on the review discussions in this PR:

- Shapely 2.0: Check performance for scalar methods/attributes #1021
- Shapely 2.0: deprecate BaseGeometry() and EmptyGeometry() ? #1022
- API: return value for the bounds of an empty geometry? #1023
- BUG: failure when importing shapely/pygeos in subprocess (outside pytest) #1024

And other outstanding issues for which I already created issues earlier:

- API: exposing CoordinateSequence in Shapely 2.0? #984
- Pickling of LinearRings #988
- API: custom classes for GEOS exceptions #991
- Shapely 2.0: support for logging ? #998
- API: read-only access to GEOS pointer (`__geom__`, `_geom`, `_ptr`) #999
